src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1955,14 +1954,23 @@ public unsafe void HandleEvents(Interop.Sys.SocketEvents events)
                 events |= Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write;
             }
 
-            if ((events & Interop.Sys.SocketEvents.Read) != 0)
+            AsyncOperation? receiveOperation =
```

Comment:
Alternatively these could be pushed as separate events to the `_eventQueue`. That would result in less `ThreadPool.UnsafeQueueUserWorkItem` calls.

Reply:
> Alternatively these could be pushed as separate events to the _eventQueue. That would result in less ThreadPool.UnsafeQueueUserWorkItem calls.

@tmds would it not double the number of enqueued work items?

Reply:
> would it not double the number of enqueued work items?

It would add an additional entry to this ConcurrentQueue vs an additional entry to the thread pool's ConcurrentQueue.

Reply:
> thread pool's ConcurrentQueue.

And the overhead of ThreadPool Enqueue.

Reply:
It might delay an IO op that became ready before others. Not sure how important it is to run them roughly in order (at least across batches of epoll notifications). I was also thinking it may be better to do the IO work on the same thread and schedule work items for the actual user callback. Would have to try some things.

Reply:
https://github.com/dotnet/runtime/pull/35330#discussion_r414004655 would also address this.

Reply:
I think a better alternative would be to process the IO on the same thread and consider scheduling the actual user callback to the thread pool if there is more than one. There are other combinations too, for example it may be beneficial overall to schedule a work item for user callbacks here and avoid some work items on the Kestrel side instead. I'm gonna leave those for future experiments.